### ISSUE TITLE
Bug: the AST type exposed by `typescript-estree` doesn't work

### ISSUE THREAD

**Issue by ericmorand** [typescript-eslint/typescript-eslint#9879]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Repro Code

Consider the following TypeScript code.

```TypeScript
import type {AST} from "@typescript-eslint/typescript-estree";

type P = AST<{
    loc: false;
    range: false;
}>;

const p = {} as P;

p.loc.end;
p.range[0];
```

### Expected Result

The `p.loc.end` expression should type check with a TS18048 error.
The `p.range[0]` expression should type check with a TS18048 error.

### Actual Result

The `p.loc.end` expression type checks successfully.
The `p.range[0]` expression type checks successfully.

### Additional Info

The `AST` type doesn't serve any purpose or doesn't work as expected.

If it is not used to give to the compiler a hint about the shape of the AST instance, then what is its purpose?
If it is supposed to give to the compiler a hint about the shape of the AST instance, then it does not work.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `8.2.0` |

**Comment by bradzacher** [comment]
Have you had a look at how the `AST` type is defined...? It's not configurable in terms of range and loc. It never has been.

https://github.com/typescript-eslint/typescript-eslint/blob/ee38b525449513ce56632aee4a7a8d40871672ee/packages/typescript-estree/src/parser.ts#L91-L93

We don't change the type based on the range and loc options because those impact the entire AST - every single AST node type would need to have those properties removed - and it's simply not possible to make that work in TS.

**Comment by ericmorand** [comment]
> It never has been

Sorry, but it is.

I can answer with the same tone: have you had a look at how the AST type is defined...? The type parameter of AST extends TSESTreeOptions which contains the loc and range properties.

**Comment by ericmorand** [comment]
By the way, can you reopen? If not, I'll create another issue asking the same question - this time being more accurate by asking why you enforce a type parameter that extends somehting that you don't need and use.

**Comment by ericmorand** [comment]
Also, I actually read the type definitions and it is made very clear that the `loc` property controls where the nodes (every single one in the AST) come with a `loc` property of not:

```ts
/**
     * Controls whether the `loc` information to each node.
     * The `loc` property is an object which contains the exact line/column the node starts/ends on.
     * This is similar to the `range` property, except it is line/column relative.
     */
    loc?: boolean;
```

Straight from you code. It seems like the parser is emitting a different shape depending on the value of the `loc` option and that you know that when writing the code since you decided to force type the returned program so that it matches the expected interface even though yu have no guarantee that it does:

https://github.com/typescript-eslint/typescript-eslint/blob/ee38b525449513ce56632aee4a7a8d40871672ee/packages/typescript-estree/src/parser.ts#L147C5-L147C27

If you are not happy to discuss your design decisions and the potential issues with your code, so be it. But please, ask another maintainer to come and discuss about the design philosophy instead of acting like nothing is wrong.

**Comment by Josh-Cena** [comment]
@ericmorand: you are filing this as a bug. However, we have neither ever made this work nor claimed that it worked, so this is definitely not a bug, it's just a non-existent API to start with.

Theoretically, it would be really cool if `loc` can have an effect on the AST shape, it's just totally not possible to do so.

**Comment by ericmorand** [comment]
> However, we have neither ever made this work nor claimed that it worked, so this is definitely not a bug, it's just a non-existent API to start with.

Fair enough. Still the question remains: why is the package exposing an AST type that expects a `TSESTreeOptions` parameter if this parameter is never used? What was the rational that drove the requirement for this parameter?

> Theoretically, it would be really cool if loc can have an effect on the AST shape, it's just totally not possible to do so.

It is.

```ts
import {AST} from "eslint";
import SourceLocation = AST.SourceLocation;

type Options = {
    loc: boolean;
}

type Base<T extends Options> = {
    loc: T["loc"] extends true ? SourceLocation : undefined;
};

type Literal<T extends Options> = Base<T>;

type Directive<T extends Options> = Base<T> & {
    type: "ExpressionStatement";
    expression: Literal<T>;
    directive: string;
};

type Program<T extends Options> = Base<T> & {
    type: "Program";
    body: Array<Directive<T>>;
};

const parse = <T extends Options>(options: T): Program<T> => {
    const loc: SourceLocation | undefined = options.loc ? {
        start: {
            column: 0,
            line: 0
        },
        end: {
            column: 0,
            line: 0
        }
    } : undefined;

    return {
        type: "Program",
        loc,
        body: {
            loc,
            type: "ExpressionStatement"
        }
    } as Program<T>;
};

const programWithLoc = parse({
    loc: true
});

programWithLoc.loc.start; // valid
programWithLoc.body[0].loc.start; // valid
programWithLoc.body[0].expression.loc.start; // valid

const programWithoutLoc = parse({
    loc: false
});

programWithoutLoc.loc.start; // TS18048: programWithoutLoc.loc is possibly undefined
programWithoutLoc.body[0].loc.start; // TS2532: Object is possibly undefined
programWithoutLoc.body[0].expression.loc.start; // TS2532: Object is possibly undefined

```

Granted, it would require that every Node type is reworked as generic, but saying "it's just totally not possible to do so" is wrong.

**Comment by Josh-Cena** [comment]
> why is the package exposing an AST type that expects a `TSESTreeOptions` parameter if this parameter is never used? What was the rational that drove the requirement for this parameter?

We accept all options but only use some of them. This ensures that you can always pass valid estree options to it and get something back, and it won't break if we change either `TSESTreeOptions` or `AST<>`. There's no guarantee that every option has to do something.

> It is.

Sure, but you see how clunky it is. I would personally not be opposed to this change, but I don't know the utility of it—after all, the point of the AST spec is to have something that works and doesn't get in the way, not to 100% reflect runtime possibilities—though that is the ultimate goal, we also have to weigh the complexity tradeoffs.

**Comment by bradzacher** [comment]
You're also only looking at just ONE AST node. There is an entire set of hundreds of AST nodes that we would have to make generic and pass this configuration option down through.

So I will revise my original comment

> We don't change the type based on the range and loc options because those impact the entire AST - every single AST node type would need to have those properties removed - ~and it's simply not possible to make that work in TS.~

It's doable but a MASSIVE burden to do so for what is ultimately a very, very rare usecase of removing range/loc. And so is not something we shall do.
